Src/VimCore/Unicode.fs

```diff
@@ -0,0 +1,2583 @@
+﻿#light
+
+namespace Vim
+
+[<RequireQualifiedAccess>]
+type internal EastAsianWidth = 
+    | FullWidth
+    | HalfWidth
+    | Wide
+    | Narrow
+    | Ambiguous
+
+    with
+
+    static member OfText text = 
+        match text with 
+        | "F" -> EastAsianWidth.FullWidth
+        | "H" -> EastAsianWidth.HalfWidth
+        | "W" -> EastAsianWidth.Wide
+        | "N" -> EastAsianWidth.Narrow
+        | "Na" -> EastAsianWidth.Narrow
+        | "A" -> EastAsianWidth.Ambiguous
+        | _ -> invalidArg "text" "Not a valid value"
+
+type internal UnicodeRangeEntry = {
+    Start: int
+    Last: int
+    Width: EastAsianWidth
+}
+
+    with
+
+    member x.Contains codePoint = codePoint >= x.Start && codePoint <= x.Last
+
+    override x.ToString() = sprintf "%d -> %d %O" x.Start x.Last x.Width
+
+type internal IntervalTreeNode
+    (
+        _entry: UnicodeRangeEntry,
+        _left: IntervalTreeNode option,
+        _right: IntervalTreeNode option
+    ) =
+
+    let _height = 1 + (max (IntervalTreeNode.GetHeight _left) (IntervalTreeNode.GetHeight _right))
+    let _count = 1 + (IntervalTreeNode.GetCount _left) + (IntervalTreeNode.GetCount _right)
+
+    new(entry: UnicodeRangeEntry) =
+        IntervalTreeNode(entry, None, None)
+
+    member x.Entry = _entry
+    member x.Left = _left
+    member x.Right = _right
+    member x.Height = _height
+    member x.Count = _count
+    member x.BalanceFactor = 
+        let leftHeight = IntervalTreeNode.GetHeight x.Left
+        let rightHeight = IntervalTreeNode.GetHeight x.Right
+        rightHeight - leftHeight
+
+    member x.WithEntry entry = IntervalTreeNode(entry, x.Left, x.Right)
+    member x.WithLeft (left: IntervalTreeNode) = IntervalTreeNode(x.Entry, Some left, x.Right)
+    member x.WithLeft (left: IntervalTreeNode option) = IntervalTreeNode(x.Entry, left, x.Right)
+    member x.WithRight (right: IntervalTreeNode) = IntervalTreeNode(x.Entry, x.Left, Some right)
+    member x.WithRight (right: IntervalTreeNode option) = IntervalTreeNode(x.Entry, x.Left, right)
+
+    static member private GetHeight (node: IntervalTreeNode option) = 
+        match node with 
+        | None -> 0
+        | Some node -> node.Height
+
+    static member private GetCount (node: IntervalTreeNode option) = 
+        match node with 
+        | None -> 0
+        | Some node -> node.Count
+
+/// A balanced internal tree for efficient lookup of unicode information
+type internal IntervalTree

```

Comment:
As a data structure enthusiast, I gotta appreciate using the right tool for the right job.